### DISCUSSION TITLE
Qbv function and Gate control list

### DISCUSSION THREAD

**Discussion by Jiang-Yu0316** [CoRE-RG/CoRE4INET#36] [Q&A]
Hello, CoRE group,

Recently, I researched the example "IEEE8021Qbv" of CoRE4INET, and came up with 4 questions, hoping to hear your thoughts.

1. In your framework, what the policy of assigning traffic to a specific queue?  based on the priority of that traffic? To be more specific, assume in node1, it can generate Q-tag traffic with priority 5, this traffic will be automatically inserted to Queue 5 or we can set up the parameter to make this traffic to a specific queue, e.g., queue 7.

2.  What does the "parameter[0]" (see below) parameter mean?
![image](https://user-images.githubusercontent.com/62637938/77508455-15f58c00-6eae-11ea-81d8-7d347419e95a.png)

3. What is the hyper-period of GCL? Does it equal to the sum of parameter “period” in GCL? E.g., a GCL is that ""C,C,C,C,C,C,C,C:0;C,C,C,C,C,C,C,o:0.00015;o,C,C,C,C,C,C,C:0.0003"", does it mean this GCL will be performed repeatly every 0.00045 sec? From the simulation result, it seems not...

4. How to measure the end-to-end delay from node1 (talker) to node2 (listener)?

Thank you in advance.
Best regards,

**Comment by Jiang-Yu0316** [comment]
Hello,
I have tested it again.
For Question 1, the traffic with priority 5 will be automatically assigned to queue5.
For Question 3, I made some mistakes, and the hyper-period equals to the sum of the parameter “period” in GCL. 
**Questions 2 and 4 are remained...**

Also, a new question occurred..
Q5：The GCL of Node1 is as below,
![image](https://user-images.githubusercontent.com/62637938/77514960-15afbd80-6ebb-11ea-8b3c-855cc2180ee9.png)

The traffic with priority 6 should be sent out from node 1 at "0.00015", "0.0006"... (send interval 500us, link-speed 100Mbps)
But, I check the log as shown below,
![image](https://user-images.githubusercontent.com/62637938/77515174-66bfb180-6ebb-11ea-8437-c7d92ce2ef03.png)

The first instance of sending traffic out is "0.00015", which maps to GCL. 
But, the second instance "0.000650163754" is not "0.0006", which against GCL.
why this happens... 
Thank you in advance
Best regards

**Comment by Musteblume** [comment]
Hello,

1 - 3: our current gcl implementations works as follows:
- gateControlList.period defines the scheduler period used for this gcl (example: 1ms)
- **.gateControlList.period sets the value for all gcls
- this period is synced between devices (default: dummySync)
- The time value in the controlList is the point in time in the period not a cumulated definition.

4:  All traffic sink apps should have a latency statistic. (In result view filter for your device and  *Latency*)

Hope this helps.
Best regards

**Comment by Jiang-Yu0316** [comment]
Hello,

**1.** 
>  **gateControlList.period defines the scheduler period used for this gcl (example: 1ms)**

If I want to set "1ms" as the period, then:
`node1.gateControlList.period="1ms"`
Is this right?

**2.**

>  **The time value in the controlList is the point in time in the period not a cumulated definition.**

then the GCL "C,C,C,C,C,C,C,C:T1; C,C,C,C,C,C,C,o:T2; ,", According to your description, this GCL means:

>          at time-instant of T1, all gate closed; 
>          at time-instant of T2, the gate of queue7 is set to open, others are kept closed
>          In addition, due to the passage of time, T2 should always be greater than T1 (because T1 and T2 is the time-instant, not a time duration)
Is my understanding right?

Thank you very much!!
Best regards,

**Comment by Musteblume** [comment]
Hi,
you are welcome!

1. You have to use a predefined period of the node scheduler:
```
**.scheduler.numPeriods = 1
**.scheduler.period[0].cycle_ticks = sec_to_tick(500us)

**.gateControlList.period = "period[0]"
```

2. Yes your understanding is right. The max of T is the length of the defined period.

Best regards

**Comment by Jiang-Yu0316** [comment]
Thank you very much. Until now all my questions have been addressed and thus I will close this issue. Thank you again.
Best regards.
